### DISCUSSION TITLE
Custom objective functions

### DISCUSSION THREAD

**Discussion by jpfeil** [online-ml/river#1284] [Ideas]
I often have several outcomes I'm trying to optimize for. Like predicting the outcome label, but I would also like to simultaneously minimize cost and time. For simple models, I would create my own custom objective function, but for more complex models, it becomes non-trivial to add a custom objective. 

I was wondering if anyone here has a solution to this using river or if river could incorporate a mechanism for easily customizing objective functions for say the adaptive random forest model. Perhaps sympy could be used to ensure the objectives derivatives?

**Comment by MaxHalford** [comment]
Hey there. I'm not sure exactly what you mean my minimizing cost and time. Can you provide more details? Usually an ML loss minimizes something that is with respect with the data, and not something extraneous such as cost and time. Is this something you've seen elsewhere in other libraries?

**Reply by jpfeil** [reply]
yeah, I think the vast majority of libraries minimize the prediction error, so usually you would have to build your own ML algorithm or edit an existing library to accept extraneous data like cost or time. Say you are writing an online advertisement recommender. You want to match the right ad to each individual but you also want to maximize profit. So you would want the ml algorithm to minimize the prediction error while also accounting for the profit from the advertisement, so you could put both pieces of data in the objective function. Otherwise, you might make good predictions but you leave some money on the table from recommending ads that don't make much in revenue.

**Reply by MaxHalford** [reply]
I think the traditional way to do that is through A/B testing and such. I don't think it's something you'd likely want to embed in your model. At the very least, it's not something you would be to do with River, sorry.
